File: mac-app/Tests/QuartzTeachersTests/MultiDestinationDeployRunnerTests.swift
import XCTest
@testable import QuartzTeachers

/// The pure decision logic a multi-destination deploy runs on: refusal
/// checks, milestone selection, and turning a finished run's outcome into
/// the sentence a teacher hears. None of this touches a real process — the
/// actual sequencing (`MultiDestinationDeployRunner.run`) needs Docker and
/// real network access, exactly like `deployAndWait()` always has, so it is
/// exercised by hand against the real app rather than here.
@MainActor
final class MultiDestinationDeployRunnerTests: XCTestCase {

    // MARK: - Refusal, checked against every configured destination

    func testRefusesOnAnAdditionalCloudflareTargetWithNoAccountID() {
        let destinations: [CourseConfiguration.DeployDestination] = [
            CourseConfiguration.DeployDestination(type: "netlify", path: ""),
            CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: ""),
        ]
        let refusal: String? = MultiDestinationDeployRunner.refusalReason(
            destinations: destinations, cloudflareAccountID: ""
        )
        XCTAssertNotNil(refusal)
        XCTAssertTrue(refusal!.contains("Account ID"))
    }

    func testRefusesOnAnAdditionalFolderTargetWithNoValidPath() {
        let destinations: [CourseConfiguration.DeployDestination] = [
            CourseConfiguration.DeployDestination(type: "netlify", path: ""),
            CourseConfiguration.DeployDestination(type: "local_folder", path: ""),
        ]
        let refusal: String? = MultiDestinationDeployRunner.refusalReason(
            destinations: destinations, cloudflareAccountID: ""
        )
        XCTAssertNotNil(refusal)
        XCTAssertTrue(refusal!.contains("folder"))
    }

    func testNoRefusalWhenEveryDestinationIsUsable() {
        let destinations: [CourseConfiguration.DeployDestination] = [
            CourseConfiguration.DeployDestination(type: "netlify", path: ""),
            CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: ""),
            CourseConfiguration.DeployDestination(type: "local_folder", path: NSTemporaryDirectory()),
        ]
        XCTAssertNil(MultiDestinationDeployRunner.refusalReason(
            destinations: destinations, cloudflareAccountID: "0123456789abcdef0123456789abcdef"
        ))
    }

    func testNetlifyNeedsNoCredentialAsEitherPrimaryOrAdditional() {
        // Netlify's token lives in Keychain, injected by the launcher —
        // there is nothing for the app itself to check.
        let destinations: [CourseConfiguration.DeployDestination] = [
            CourseConfiguration.DeployDestination(type: "local_folder", path: NSTemporaryDirectory()),
            CourseConfiguration.DeployDestination(type: "netlify", path: ""),
        ]
        XCTAssertNil(MultiDestinationDeployRunner.refusalReason(
            destinations: destinations, cloudflareAccountID: ""
        ))
    }

    // MARK: - Milestones stay destination-accurate and never mention the wrong host

    func testDeployOnlyMilestonesNeverMentionNetlifyForOtherDestinations() {
        for milestone in MultiDestinationDeployRunner.deployOnlyMilestones(forDestinationType: "cloudflare_pages") {
            XCTAssertFalse(milestone.label.contains("Netlify"))
        }
        for milestone in MultiDestinationDeployRunner.deployOnlyMilestones(forDestinationType: "local_folder") {
            XCTAssertFalse(milestone.label.contains("Netlify"))
        }
    }

    func testBuildAndDeployMilestonesMatchDeployOnlyMilestonesForTheSameType() {
        // Every leg after the first always uses the deploy-only list,
        // because the build (if one happened) already ran on leg one —
        // both lists still have to agree about which HOST they name.
        for type in CourseConfiguration.knownDeployTargetTypes {
            let buildAndDeploy: [TaskMilestone] = MultiDestinationDeployRunner.buildAndDeployMilestones(forDestinationType: type)
            let deployOnly: [TaskMilestone] = MultiDestinationDeployRunner.deployOnlyMilestones(forDestinationType: type)
            XCTAssertEqual(buildAndDeploy.last?.marker, deployOnly.last?.marker,
                           "Both must end on the same 'finished' marker for \(type)")
        }
    }

    // MARK: - Joining destination names

    func testJoinedWithAndHandlesEveryCount() {
        XCTAssertEqual(MultiDestinationDeployRunner.joinedWithAnd([]), "")
        XCTAssertEqual(MultiDestinationDeployRunner.joinedWithAnd(["Netlify"]), "Netlify")
        XCTAssertEqual(MultiDestinationDeployRunner.joinedWithAnd(["Netlify", "Cloudflare Pages"]),
                       "Netlify and Cloudflare Pages")
        XCTAssertEqual(
            MultiDestinationDeployRunner.joinedWithAnd(["Netlify", "Cloudflare Pages", "/Users/teacher/Sites"]),
            "Netlify, Cloudflare Pages and /Users/teacher/Sites"
        )
    }

    // MARK: - Turning an outcome into the sentence a teacher hears

    /// A course with exactly one destination — the overwhelming majority —
    /// must use the UNCHANGED single-destination wording, word for word,
    /// whatever `outcome` says. Multi-destination sentences must never
    /// appear for a teacher who never opted into redundancy.
    func testASingleDestinationCourseAlwaysUsesTheUnchangedWording() {
        let succeeded: Outcome = Outcome(anySucceeded: true, failedDestinations: [])
        let succeededResult: AssistSiteWorkResult = MultiDestinationDeployRunner.result(
            course: "ICS3U", section: "1", destinationCount: 1, outcome: succeeded
        )
        XCTAssertEqual(succeededResult.message, AssistWording.deployed(course: "ICS3U", section: "1"))

        let failed: Outcome = Outcome(anySucceeded: false, failedDestinations: [
            CourseConfiguration.DeployDestination(type: "netlify", path: ""),
        ])
        let failedResult: AssistSiteWorkResult = MultiDestinationDeployRunner.result(
            course: "ICS3U", section: "1", destinationCount: 1, outcome: failed
        )
        XCTAssertEqual(failedResult.message, AssistWording.deployDidNotFinish(course: "ICS3U", section: "1"))
    }

    func testAllDestinationsSucceeding() {
        let outcome: Outcome = Outcome(anySucceeded: true, failedDestinations: [])
        let result: AssistSiteWorkResult = MultiDestinationDeployRunner.result(
            course: "ICS3U", section: "1", destinationCount: 2, outcome: outcome
        )
        XCTAssertTrue(result.succeeded)
        XCTAssertEqual(result.message, AssistWording.deployedToMultipleDestinations(
            course: "ICS3U", section: "1", destinationCount: 2
        ))
    }

    func testSomeDestinationsFailingNamesWhichOnes() {
        let outcome: Outcome = Outcome(anySucceeded: true, failedDestinations: [
            CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: ""),
        ])
        let result: AssistSiteWorkResult = MultiDestinationDeployRunner.result(
            course: "ICS3U", section: "1", destinationCount: 2, outcome: outcome
        )
        XCTAssertFalse(result.succeeded)
        XCTAssertEqual(result.message, AssistWording.deployPartiallySucceeded(
            course: "ICS3U", section: "1", failedDestinations: "Cloudflare Pages"
        ))
    }

    func testEveryDestinationFailing() {
        let outcome: Outcome = Outcome(anySucceeded: false, failedDestinations: [
            CourseConfiguration.DeployDestination(type: "netlify", path: ""),
            CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: ""),
        ])
        let result: AssistSiteWorkResult = MultiDestinationDeployRunner.result(
            course: "ICS3U", section: "1", destinationCount: 2, outcome: outcome
        )
        XCTAssertFalse(result.succeeded)
        XCTAssertEqual(result.message, AssistWording.deployToMultipleDestinationsDidNotFinish(course: "ICS3U", section: "1"))
    }

    // MARK: - The outcome computed property, against hand-built legs

    @MainActor
    func testOutcomeCountsOnlyFinishedLegsAndExcludesBuildFailures() {
        let runner: MultiDestinationDeployRunner = MultiDestinationDeployRunner()
        var netlifyLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "netlify", path: "")
        )
        netlifyLeg.isFinished = true
        netlifyLeg.succeeded = true

        var cloudflareLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: "")
        )
        cloudflareLeg.isFinished = true
        cloudflareLeg.succeeded = false

        // Never actually reached — a cancel or an earlier build failure
        // stopped the run before this leg's turn came.
        let folderLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "local_folder", path: NSTemporaryDirectory())
        )

        runner.legs = [netlifyLeg, cloudflareLeg, folderLeg]

        let outcome: MultiDestinationDeployRunner.Outcome = runner.outcome
        XCTAssertTrue(outcome.anySucceeded)
        XCTAssertEqual(outcome.failedDestinations, [
            CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: ""),
        ])
        XCTAssertFalse(outcome.allSucceeded)
    }

    @MainActor
    func testABuildFailureIsNeverCountedAsAFailedDestination() {
        // A build failure means no destination was ever actually
        // attempted — it must not show up in "failedDestinations" (which
        // would wrongly claim Netlify itself rejected the upload).
        let runner: MultiDestinationDeployRunner = MultiDestinationDeployRunner()
        var leg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "netlify", path: "")
        )
        leg.isFinished = true
        leg.buildFailed = true
        runner.legs = [leg]

        let outcome: MultiDestinationDeployRunner.Outcome = runner.outcome
        XCTAssertFalse(outcome.anySucceeded)
        XCTAssertTrue(outcome.failedDestinations.isEmpty)
    }

    @MainActor
    func testActiveRunnerFallsBackSafelyWithNoLegs() {
        let runner: MultiDestinationDeployRunner = MultiDestinationDeployRunner()
        // Must not crash — a fresh orchestrator with nothing run yet.
        _ = runner.activeRunner
        XCTAssertFalse(runner.hasAnyOutput)
    }

    // MARK: - DeployDestinationLinks.succeeded(from:) — the fix for
    // "only Cloudflare, the second deploy target, is visible" once a
    // multi-destination deploy finishes.

    @MainActor
    func testSucceededKeepsOnlyFinishedAndSucceededLegs() {
        var netlifyLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "netlify", path: "")
        )
        netlifyLeg.isFinished = true
        netlifyLeg.succeeded = true

        var cloudflareLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: "")
        )
        cloudflareLeg.isFinished = true
        cloudflareLeg.succeeded = false

        // Never reached — a cancel or an earlier build failure stopped the
        // run before this leg's turn came.
        let folderLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "local_folder", path: NSTemporaryDirectory())
        )

        let succeeded: [MultiDestinationDeployRunner.Leg] = DeployDestinationLinks.succeeded(
            from: [netlifyLeg, cloudflareLeg, folderLeg]
        )
        XCTAssertEqual(destinationTypes(of: succeeded), ["netlify"])
    }

    @MainActor
    func testSucceededListsEveryDestinationThatActuallyPublished() {
        // The exact bug reported: after a successful deploy to BOTH
        // Netlify and Cloudflare, only the last one to run was visible —
        // this is the property that must hold once both have finished.
        var netlifyLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "netlify", path: "")
        )
        netlifyLeg.isFinished = true
        netlifyLeg.succeeded = true

        var cloudflareLeg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: "")
        )
        cloudflareLeg.isFinished = true
        cloudflareLeg.succeeded = true

        let succeeded: [MultiDestinationDeployRunner.Leg] = DeployDestinationLinks.succeeded(
            from: [netlifyLeg, cloudflareLeg]
        )
        XCTAssertEqual(destinationTypes(of: succeeded), ["netlify", "cloudflare_pages"])
    }
}

/// Named after what it produces, not how — used only to keep the
/// assertions above readable without reaching for `.map`.
@MainActor
private func destinationTypes(of legs: [MultiDestinationDeployRunner.Leg]) -> [String] {
    var types: [String] = []
    for leg in legs {
        types.append(leg.destination.type)
    }
    return types
}

private typealias Outcome = MultiDestinationDeployRunner.Outcome

File: windows-app/Plantoir.Tests/MultiDestinationDeployRunnerTests.cs
using Plantoir.Core.Models;
using Plantoir.Core.Scripting;
using Xunit;

namespace Plantoir.Tests;

/// <summary>
/// Mirrors the mac's MultiDestinationDeployRunnerTests.swift — the pieces
/// testable without actually spawning deploy.ps1 (refusal reasoning,
/// milestone selection, outcome bookkeeping, and wording selection). The
/// sequencing itself (RunAsync) is exercised end to end by driving a real
/// deploy, the same way the mac's own sequencing is.
/// </summary>
public class MultiDestinationDeployRunnerTests
{
    private static readonly CourseConfiguration.DeployDestination Netlify = new("netlify", "");
    private static readonly CourseConfiguration.DeployDestination Cloudflare = new("cloudflare_pages", "");
    private static readonly string ValidFolderPath = Directory.CreateTempSubdirectory("multidest-runner-tests").FullName;
    private static readonly CourseConfiguration.DeployDestination Folder = new("local_folder", ValidFolderPath);
    private static readonly CourseConfiguration.DeployDestination BadFolder = new("local_folder", "");

    [Fact]
    public void RefusesOnAnAdditionalCloudflareTargetWithNoAccountID()
    {
        string? refusal = MultiDestinationDeployRunner.RefusalReason(
            new[] { Netlify, Cloudflare }, cloudflareAccountID: "");
        Assert.NotNull(refusal);
        Assert.Contains("Account ID", refusal);
    }

    [Fact]
    public void RefusesOnAnAdditionalFolderTargetWithNoValidPath()
    {
        string? refusal = MultiDestinationDeployRunner.RefusalReason(
            new[] { Netlify, BadFolder }, cloudflareAccountID: "0123456789abcdef0123456789abcdef");
        Assert.NotNull(refusal);
        Assert.Contains("folder", refusal, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void NoRefusalWhenEveryDestinationIsUsable()
    {
        string? refusal = MultiDestinationDeployRunner.RefusalReason(
            new[] { Netlify, Cloudflare, Folder }, cloudflareAccountID: "0123456789abcdef0123456789abcdef");
        Assert.Null(refusal);
    }

    [Fact]
    public void NetlifyNeedsNoCredentialAsEitherPrimaryOrAdditional()
    {
        Assert.Null(MultiDestinationDeployRunner.RefusalReason(new[] { Netlify }, ""));
        Assert.Null(MultiDestinationDeployRunner.RefusalReason(new[] { Cloudflare, Netlify }, "0123456789abcdef0123456789abcdef"));
    }

    [Fact]
    public void DeployOnlyMilestonesDifferPerDestinationType()
    {
        Assert.NotEqual(MultiDestinationDeployRunner.DeployOnlyMilestones("netlify"),
                         MultiDestinationDeployRunner.DeployOnlyMilestones("cloudflare_pages"));
        Assert.Equal(TaskMilestones.Deploy, MultiDestinationDeployRunner.DeployOnlyMilestones("netlify"));
        Assert.Equal(TaskMilestones.DeployToCloudflare, MultiDestinationDeployRunner.DeployOnlyMilestones("cloudflare_pages"));
        Assert.Equal(TaskMilestones.DeployToFolder, MultiDestinationDeployRunner.DeployOnlyMilestones("local_folder"));
    }

    [Fact]
    public void BuildAndDeployMilestonesMatchTheSingleDestinationLists()
    {
        Assert.Equal(TaskMilestones.BuildAndDeploy, MultiDestinationDeployRunner.BuildAndDeployMilestones("netlify"));
        Assert.Equal(TaskMilestones.BuildAndDeployToCloudflare, MultiDestinationDeployRunner.BuildAndDeployMilestones("cloudflare_pages"));
        Assert.Equal(TaskMilestones.BuildAndDeployToFolder, MultiDestinationDeployRunner.BuildAndDeployMilestones("local_folder"));
    }

    [Theory]
    [InlineData(new string[] { }, "")]
    [InlineData(new[] { "Cloudflare Pages" }, "Cloudflare Pages")]
    [InlineData(new[] { "Cloudflare Pages", "your folder" }, "Cloudflare Pages and your folder")]
    [InlineData(new[] { "Netlify", "Cloudflare Pages", "your folder" }, "Netlify, Cloudflare Pages and your folder")]
    public void JoinedWithAndHandlesEveryCount(string[] items, string expected) =>
        Assert.Equal(expected, MultiDestinationDeployRunner.JoinedWithAnd(items));

    [Fact]
    public void ASingleDestinationCourseAlwaysUsesTheUnchangedWording()
    {
        var outcome = new MultiDestinationDeployRunner.Outcome(true, Array.Empty<CourseConfiguration.DeployDestination>());
        var result = MultiDestinationDeployRunner.Result("ICS3U", "1", destinationCount: 1, outcome);
        Assert.True(result.Succeeded);
        Assert.Equal(Plantoir.Core.Assist.AssistWording.Deployed("ICS3U", "1"), result.Message);
    }

    [Fact]
    public void AllDestinationsSucceeding()
    {
        var outcome = new MultiDestinationDeployRunner.Outcome(true, Array.Empty<CourseConfiguration.DeployDestination>());
        var result = MultiDestinationDeployRunner.Result("ICS3U", "1", destinationCount: 2, outcome);
        Assert.True(result.Succeeded);
        Assert.Contains("all 2 destinations", result.Message);
    }

    [Fact]
    public void SomeDestinationsFailingNamesWhichOnes()
    {
        var outcome = new MultiDestinationDeployRunner.Outcome(true, new[] { Cloudflare });
        var result = MultiDestinationDeployRunner.Result("ICS3U", "1", destinationCount: 2, outcome);
        Assert.False(result.Succeeded);
        Assert.Contains("Cloudflare Pages", result.Message);
    }

    [Fact]
    public void EveryDestinationFailing()
    {
        var outcome = new MultiDestinationDeployRunner.Outcome(false, new[] { Netlify, Cloudflare });
        var result = MultiDestinationDeployRunner.Result("ICS3U", "1", destinationCount: 2, outcome);
        Assert.False(result.Succeeded);
        Assert.Contains("did not finish, on any of its destinations", result.Message);
    }

    [Fact]
    public void ActiveRunnerFallsBackSafelyWithNoLegs()
    {
        var runner = new MultiDestinationDeployRunner();
        Assert.NotNull(runner.ActiveRunner);
        Assert.False(runner.ActiveRunner.IsRunning);
    }

    [Fact]
    public void OutcomeAllSucceededRequiresAtLeastOneSuccessAndNoFailures()
    {
        Assert.True(new MultiDestinationDeployRunner.Outcome(true, Array.Empty<CourseConfiguration.DeployDestination>()).AllSucceeded);
        Assert.False(new MultiDestinationDeployRunner.Outcome(false, Array.Empty<CourseConfiguration.DeployDestination>()).AllSucceeded);
        Assert.False(new MultiDestinationDeployRunner.Outcome(true, new[] { Netlify }).AllSucceeded);
    }
}
